[RateLimiter] Add `RateLimiterBuilder`

kbond opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix https://github.com/symfony/symfony/issues/51401#issuecomment-2754958638
| License       | MIT

Configuring a rate limiter today means one `framework.rate_limiter.limiters` entry per limiter, each injected individually. When a service needs many one-off limiters, or when the limits are only known at runtime, that does not fit. `RateLimiterBuilder` closes the gap: inject one service and build limiter factories on the fly.

```php
use Symfony\Component\RateLimiter\RateLimiterBuilder;

class ApiController extends AbstractController
{
    public function registerUser(Request $request, RateLimiterBuilder $limiterBuilder): Response
    {
        $limiter = $limiterBuilder
            ->slidingWindow(id: 'register', limit: 10, interval: '10 minutes')
            ->create($request->getClientIp());

        // ...
    }
}
```

Every method returns a `RateLimiterFactoryInterface`, so `create($key)` stays the single entry point and the per-key lock comes from the factory, exactly as for configured limiters:

- `slidingWindow(string $id, int $limit, \DateInterval|string $interval)`
- `fixedWindow(string $id, int $limit, \DateInterval|string $interval, \DateTimeInterface|string|null $anchorAt = null)`, where `$anchorAt` aligns the window to a calendar instead of the first hit
- `tokenBucket(string $id, int $limit, \DateInterval|string $interval, int $amount = 1)`
- `compound(RateLimiterFactoryInterface ...$factories)`
- `noop()`, a factory of `NoLimiter`, to disable limiting conditionally

String intervals accept the same units as the configuration (`'10 minutes'`, ...). `\DateInterval` is accepted too; that also works for the `interval` and `rate.interval` options of `RateLimiterFactory` itself now.

FrameworkBundle registers the `limiter_builder` service when the RateLimiter component is installed and `framework.rate_limiter` is enabled, autowirable by class name. Configuration, all optional:

```yaml
framework:
    rate_limiter:
        builder:
            lock_factory: auto      # "auto" uses lock.factory when framework.lock is enabled; null disables locking
            cache_pool: cache.rate_limiter
            storage_service: null   # a StorageInterface service id; takes precedence over cache_pool
```

Points for the documentation:
- the `$id` is the storage namespace: two limiters built with the same id and storage share their state, which also allows sharing state with a configured limiter on purpose;
- state lives in the `cache.rate_limiter` pool by default, like configured limiters;
- in the shorthand syntax (`rate_limiter: { name: {...} }`), `builder` is a reserved key; a limiter named `builder` needs the explicit `limiters:` form.

kbond committed:
[RateLimiter] Add `RateLimiterBuilder`

alexandre-daubois commented on src/Symfony/Component/RateLimiter/RateLimiterBuilder.php:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * @param string               $id       Unique identifier for the rate limiter
+     * @param int                  $limit    Maximum allowed hits for the passed interval
+     * @param \DateInterval|string $interval If string, must be a number followed by "second",
+     *                                       "minute", "hour", "day", "week" or "month" (or their
+     *                                       plural equivalent)
```
I think that we tend to write something like "Any valid date interval string" and maybe link to the right PHP doc page?

kbond replied:
I copied the note from here: https://github.com/symfony/symfony/blob/561d4a086c13bde50a16e0f5408582d2e1d85897/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L2526

It's a little custom as we don't want the + included

alexandre-daubois replied:
Alright then 🙂

stof commented on src/Symfony/Component/RateLimiter/RateLimiterBuilder.php:
```diff
@@ -0,0 +1,102 @@
+     *                                       "minute", "hour", "day", "week" or "month" (or their
+     *                                       plural equivalent)
+     */
+    public function fixedWindow(string $id, int $limit, \DateInterval|string $interval): FixedWindowLimiter
```
should we use our implementation classes as return types for those methods or should we use LimiterInterface ?

kbond replied:
I've made this change. I don't have strong opinion but no harm in returning the interface.

kbond committed:
frameworkbundle configuration/tests

kbond committed:
changelog

kbond committed:
test

kbond committed:
fix tests

kaznovac commented on src/Symfony/Component/RateLimiter/RateLimiterBuilder.php:
```diff
@@ -0,0 +1,102 @@
+        return new CompoundLimiter($limiters);
+    }
+
+    public function noop(): LimiterInterface
```
```suggestion
    public function noLimit(): LimiterInterface
```

the policy name in config for `NoLimiter` is `no_limit` - it's nice to be consistent for discoverability purposes (personally, I like the `noop` name better)

kbond replied:
Hmm, yeah I really like `noop` too - especially since the other methods don't include "Limit" in their name.

Let's see what others think.

nicolas-grekas replied:
Settling this so the thread does not stay open: keeping `noop()`.

The four other methods name a rate limiting policy, so following the config vocabulary is right for them. This one is different in kind: it does not select a policy, it switches limiting off, and `noop` states that directly at the call site. It is also the shorter name in a five-method class where nothing can be missed, and it matches @kbond's preference above.

kaznovac commented on src/Symfony/Component/RateLimiter/RateLimiterBuilder.php:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * @internal
+     */
+    public static function intervalNormalizer(\DateInterval|string $interval): \DateInterval
```
this helper function just does not feel that it should be in this class; especially as it is public, and this class is intended as a public API

better move it to `Util` namespace
e.g. how `FormComponent` does it https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/Form/Util/StringUtil.php

and test for it would be nice

kbond replied:
Ah, there was already a `TimeUtil`, I moved there.

> and test for it would be nice

This function tested in the existing `RateLimiterFactoryTest`.

kaznovac commented on src/Symfony/Component/RateLimiter/RateLimiterBuilder.php:
```diff
@@ -0,0 +1,102 @@
+     *                                       "minute", "hour", "day", "week" or "month" (or their
+     *                                       plural equivalent)
+     */
+    public function slidingWindow(string $id, int $limit, \DateInterval|string $interval): LimiterInterface
```
```suggestion
    public function slidingWindow(string $id, int $limit, \DateInterval|string $interval): SlidingWindowLimiter
```

this is a bit a nitpick... but I like type correctness/safety where possible

this function is intended to create a `slidingWindow` or its specialization (yes the limiter classes are final so specialization is out-of-reach; but his might change in the future)

generalization here potentially allows `slidingWindow` to return any other Limiter policy in other `RateLimiterBuilder` specializations (consider marking this class as `final`) which is undesirable/unexpected... and might require upstream casting of return value if the concrete implementation method/property is required (currently not the case - all Limiter implementation have the same interface)

stof replied:
Using the implementation class name in the return type prevents us from doing any refactoring in the future (like returning a decorated implementation for instance) as widening a return type is a BC break.
For instance, we could imagine that in the future the returned limiters could be decorated with a traceable limiter in dev to be integrated with the profiler.

And having the same interface for all limiter implementations is the whole point of having a LimiterInterface as the main type in the component.

kaznovac replied:
TL;DR
I'd suggest making this `RateLimiterBuilder` class `final`

---

> Using the implementation class name in the return type prevents us from doing any refactoring in the future (like returning a decorated implementation for instance) as widening a return type is a BC break.

I see it as a way to prevent an unsafe change; when the type changes this communicates to upstream devs that they need to update their code.
Yes, changing this type is an explicit BC break, and that is actually good - it can be detected with static analyzers and in runtime type checking; using the interface type allows contributions that potentially introduce unexpected behavior in the runtime. Interface approach leaves the function name as the sole 'contract' - more of a phpDoc hint.

> For instance, we could imagine that in the future the returned limiters could be decorated with a traceable limiter in dev to be integrated with the profiler.

This is still possible to do with expected implementation classes, but requires more work. It would require proxy class code generation (like Doctrine does, complicated), or creating tracing specializations of each limiter policy (does not scale well), or use of some AOP library (slow)...

> And having the same interface for all limiter implementations is the whole point of having a LimiterInterface as the main type in the component.

Interface is preferably ok for https://github.com/symfony/symfony/blob/8b44a51e619301166d9d5c6afb02aa0f9c78be5e/src/Symfony/Component/RateLimiter/RateLimiterFactory.php#L42 as developer cannot make assumption on what concrete instance will be returned; arguably this is not the case for the `RateLimiterBuilder`'s methods.

---

The contract for the methods of this class is type checked within this test https://github.com/symfony/symfony/blob/eae6a0d205b5a43294e5aac94dd7d7bbdbb21d13/src/Symfony/Component/RateLimiter/Tests/RateLimiterBuilderTest.php#L27-L36
which is good, but this test cannot guarantee types returned by the methods of classes derived from the `RateLimiterBuilder`

nicolas-grekas replied:
This would go against the IoC principle, which we strive for. Symfony is a piece of reusable and extensible software, it has to provide flexibility and face the possibility that people will do silly things.

Closing the implementation like you're suggesting is on the contrary building on the idea that we should prevent end users from doing those silly things, at the cost of also preventing power users from doing what they want to achieve.

That's not the approach we have in this project as a whole. We always favor SOLID principles because they're empowering end users (nothing prevents silly things anyway.)

stof replied:
> Yes, changing this type is an explicit BC break, and that is actually good

No it is not good, because it means it is actually forbidden to change this (due to our BC policy) and so we would not be able to add support for traceable limiters in the future (duplicating traceable logic to have implementations done using inheritance instead of composition is a no-go. We did things like that in the early days of Symfony and it was a maintenance nightmare)

kaznovac replied:
with all due respect

> This would go against the IoC principle, which we strive for. ...

IoC, here, is possible _because_ this class does not follow `SOLID` principles:
`S` - breached - multiple classes instantiated
`O[C]` - not enforced - not open for extension (not implemented, do not confuse with `class extends`), not closed (this is `class extends`, and it is not forbidden)
`L` - not enforced - interface for return type makes breaching this one easy
`I` - not implemented
`D` - ok - made possible by compromises above

> Closing the implementation like you're suggesting is ...

... in the favor of `C` in `O[C]` principle of `SOLID`

> That's not the approach we have in this project as a whole. We always favor SOLID principles because they're empowering end users (nothing prevents silly things anyway.)

I agree, `SOLID` is a great design guide to follow (albeit class explosion and discovery), but this is a utility class and should probably be spared from _'power'-future-proofing_.

I strongly believe that the architecture should favor **both** empowerment of users and prevention of invalid state (a.k.a. foot-guns or even CVEs)

---

> (duplicating traceable logic to have implementations done using inheritance instead of composition is a no-go. We did things like that in the early days of Symfony and it was a maintenance nightmare)

This, however, is quite SOLID argument ;)

kaznovac commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/PhpFrameworkExtensionTest.php:
```diff
@@ -290,4 +291,50 @@ public function testRateLimiterIsTagged()
+
+        $container = $this->createContainerFromClosure(function (ContainerBuilder $container) {
+            $container->loadFromExtension('framework', [
+                'annotations' => false,
+                'http_method_override' => false,
+                'handle_all_throwables' => true,
+                'php_errors' => ['log' => true],
```
```suggestion
```

these can be removed as they are default in the target 7.x symfony

kbond committed:
review

nicolas-grekas commented:
I'd prefer a solution that doesn't create a new sibling concept that's hard to grasp vs "factory".
Can't we have something like a wither API?

kbond commented:
> Can't we have something like a wither API on the factory class?

Can you describe what you're thinking here a bit more? Feels like this ship has sailed as we now have 2 `RateLimiterFactoryInterface` implementations.

kaznovac commented:
maybe something like this:

```php
$factory->createBuilder(): RateLimiterBuilderInterface

interface RateLimiterBuilderInterface {
  function setLock(?LockInterface $lock): self; /** if desired to allow changing of factory provided services */
  function setStorage(StorageInterface $storage): self;

  function setId(string $id): self; /** or better setName */
  /** @param 'fixed_window'|'token_bucket'|'sliding_window'|'no_limit' $policy */
  function setPolicy(string $policy): self;

  function setLimit(int $limit): self;
  function setInterval(string $interval): self;
  function setRate(string $interval, int $amount): self; /** these can be noop where not-applicable to policy */

  function reset(): self; /** optional */

  function build(): LimiterInterface;
}
```

If you want to follow dogmatic Builder pattern - for each Limiter class a dedicated builder class should be crated; strenuous, but allows nice strategy pattern in the new factory's `createBuilder(string $policy): RateLimiterBuilderInterface`
(this aproach fully satisfies `O` - `open` for extension - new builders can be registered per policy, and factory is `closed` for modification as it's explicitly `final`; and typehint it is through phpstan conditional types)


Regarding `CompoundLimiterBuilder`, it can have `addLimiter(LimiterInterface $limiter): self` and just wrap all provided limiters in adequate `CompoundLimiter` instance on `build`.

nicolas-grekas reviewed (commented):
Sorry for the long silence, and for not answering when you asked me to describe what I meant.

Two things, one of them a blocker.

The lock argument is broken as wired. The constructor takes `?LockInterface $lock`, but `FrameworkExtension` injects `lock.factory`, and `LockFactory` does not implement `LockInterface`, so the service cannot be constructed once `framework.lock` is enabled. `testRateLimiterBuilderDefaultWithLock()` passes because it asserts on the definition rather than instantiating it. Beyond the type, a single lock shared by every limiter the builder creates is wrong anyway: `RateLimiterFactory::create()` builds one per id with `$this->lockFactory?->createLock($id)`.

On the design, you were right that the ship has sailed for a wither on the interface, `CompoundRateLimiterFactory` could not implement it and it would break third-party implementations. What removes my "sibling concept" objection instead is having the methods return a `RateLimiterFactory` rather than a `LimiterInterface`. Then nothing new is introduced, the builder is just a typed way to construct the factory you already know, `->create($key)` stays the single entry point, and the per-id lock comes for free because the factory does it. `TimeUtil::intervalNormalizer()` already accepts `\DateInterval|string`, so it composes; `RateLimiterFactory`'s `interval` option needs `allowedTypes('string', \DateInterval::class)`.

So: constructor takes `?LockFactory`, methods return `RateLimiterFactory`, and the rest stands. The rebase is six files, mostly `Configuration`, `FrameworkExtension` and `rate_limiter.php`; note `arrayNode('limiters')` gained a second argument and `TimeUtil` must carry the `(string) time()` fix that landed on 8.2 since.

kbond committed:
[RateLimiter] Add `RateLimiterBuilder`

Adds a builder that constructs `RateLimiterFactory` instances from a shared storage
and lock factory, so limiters can be defined in PHP without going through
`framework.rate_limiter.limiters`. Each method returns a factory, keeping
`->create($key)` as the single entry point and locks created per id. The factory's
`interval` option now also accepts a `\DateInterval`.

Registered as `limiter_builder` and configurable under `framework.rate_limiter.builder`.

kbond commented:
Thanks for the review @nicolas-grekas. I've made the changes you asked for!

nicolas-grekas commented:
Thank you @kbond, the rework addresses everything. All findings from the previous round are verified fixed; `testLockIsCreatedPerKey` now covers the lock path at the right layer.

Prep pushed on top: rebased onto the current 8.2 tip (one CHANGELOG conflict), added the missing RateLimiter changelog line for `\DateInterval` being accepted by `RateLimiterFactory`'s `interval`/`rate.interval` options, and refreshed the description. Ready on my side once CI is green.

nicolas-grekas commented:
Thank you @kbond.
